[Cocoa] Create "NonVisualMode" to avoid visual operations for embedders that don't need it

litherum opened the pull request:
#### fe6185621fb591bc747d1859c0ec216920a8fadc
<pre>
[Cocoa] Create &quot;NonVisualMode&quot; to avoid visual operations for embedders that don&apos;t need it
<a href="https://bugs.webkit.org/show_bug.cgi?id=257948">https://bugs.webkit.org/show_bug.cgi?id=257948</a>
rdar://110625441

Reviewed by NOBODY (OOPS!).

This is a mode for embedders who don&apos;t actually visually look at the web view - e.g.
embedders who only interact with the web view programmatically. For those embedders,
there&apos;s no point actually painting anything, since no one will actually view it.
In fact, painting can actually be harmful, because it takes time to do so. If we can
make WebKit faster by just avoiding painting (which is unnecessary anyway for these
embedders) then the performance improvements could be significantly valuable.

Traditionally, the way to have a non-visual mode was to just not parent the WKWebView.
However, this has a bunch of negative repercussions like throttling timers, etc.
This new mode is intended to be used with a parented WKWebView - it just avoids
painting / compositing into it.

This patch just hooks up this setting to painting operations, but in the future it
could/should be hooked up to compositing, IOSurface creation etc. This patch starts
small, though.

Beware that painting triggers full layout. When programmatically interacting with
the page, we *should* perform enough layouts to answer whatever question we are
being asked, but it&apos;s totally possible that we missed some spots, but never caught
it because they were being masked by the fact that we&apos;re painting all the time. So
it&apos;s possible that, by avoiding painting, it makes it more likely for embedders to
hit preexisting missing layout bugs, if they exist.

Tests: fast/dom/non-visual-mode.html
       fast/images/non-visual-mode.html
       fast/text/non-visual-mode.html
       fast/transforms/non-visual-mode.html

* LayoutTests/TestExpectations:
* LayoutTests/fast/dom/non-visual-mode-expected-mismatch.html: Added.
* LayoutTests/fast/dom/non-visual-mode.html: Added.
* LayoutTests/fast/images/non-visual-mode-expected-mismatch.html: Added.
* LayoutTests/fast/images/non-visual-mode.html: Added.
* LayoutTests/fast/text/non-visual-mode-expected-mismatch.html: Added.
* LayoutTests/fast/text/non-visual-mode.html: Added.
* LayoutTests/fast/transforms/non-visual-mode-expected-mismatch.html: Added.
* LayoutTests/fast/transforms/non-visual-mode.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::GraphicsLayer):
(WebCore::GraphicsLayer::paintGraphicsLayerContents):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::nonVisualModeEnabled const):
(WebCore::GraphicsLayer::setNonVisualModeEnabled):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences _setNonVisualModeEnabled:]):
(-[WKPreferences _nonVisualModeEnabled]):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::createGraphicsLayer):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe6185621fb591bc747d1859c0ec216920a8fadc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18413 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18751 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20262 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17240 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22057 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18900 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19144 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18637 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18826 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16038 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21141 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16049 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23274 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/15964 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17069 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16965 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21161 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/17701 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17540 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14880 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/21773 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16621 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5327 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20988 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/23013 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17379 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5175 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/GraphicsLayer.h:
```diff
@@ -797,6 +800,7 @@ class GraphicsLayer : public RefCounted<GraphicsLayer> {
     bool m_isDescendentOfSeparatedPortal : 1;
 #endif
 #endif
+    bool m_headlessModeEnabled : 1;
```
This should not be state on every GraphicsLayer. Instead, extend the `GraphicsLayerClient` protocol.

smfr commented on Source/WebCore/platform/graphics/GraphicsLayer.cpp:
```diff
@@ -575,7 +576,8 @@ void GraphicsLayer::paintGraphicsLayerContents(GraphicsContext& context, const F
     }
 
-    client().paintContents(this, context, clipRect, layerPaintBehavior);
+    if (!m_headlessModeEnabled)
```
This just stops painting. You probably want to also avoid backing store allocation. I think the best way to do that is to make every layer behave as if `m_drawsContent` is false if the clients says to.

litherum committed:
[Cocoa] Create "NonVisualMode" to avoid visual operations for embedders that don't need it
https://bugs.webkit.org/show_bug.cgi?id=257948
rdar://110625441

Reviewed by NOBODY (OOPS!).

This is a mode for embedders who don't actually visually look at the web view - e.g.
embedders who only interact with the web view programmatically. For those embedders,
there's no point actually painting anything, since no one will actually view it.
In fact, painting can actually be harmful, because it takes time to do so. If we can
make WebKit faster by just avoiding painting (which is unnecessary anyway for these
embedders) then the performance improvements could be significantly valuable.

Traditionally, the way to have a non-visual mode was to just not parent the WKWebView.
However, this has a bunch of negative repercussions like throttling timers, etc.
This new mode is intended to be used with a parented WKWebView - it just avoids
painting / compositing into it.

This patch just hooks up this setting to painting operations, but in the future it
could/should be hooked up to compositing, IOSurface creation etc. This patch starts
small, though.

Beware that painting triggers full layout. When programmatically interacting with
the page, we *should* perform enough layouts to answer whatever question we are
being asked, but it's totally possible that we missed some spots, but never caught
it because they were being masked by the fact that we're painting all the time. So
it's possible that, by avoiding painting, it makes it more likely for embedders to
hit preexisting missing layout bugs, if they exist.

Tests: fast/dom/non-visual-mode.html
       fast/images/non-visual-mode.html
       fast/text/non-visual-mode.html
       fast/transforms/non-visual-mode.html

* LayoutTests/TestExpectations:
* LayoutTests/fast/dom/non-visual-mode-expected-mismatch.html: Added.
* LayoutTests/fast/dom/non-visual-mode.html: Added.
* LayoutTests/fast/images/non-visual-mode-expected-mismatch.html: Added.
* LayoutTests/fast/images/non-visual-mode.html: Added.
* LayoutTests/fast/text/non-visual-mode-expected-mismatch.html: Added.
* LayoutTests/fast/text/non-visual-mode.html: Added.
* LayoutTests/fast/transforms/non-visual-mode-expected-mismatch.html: Added.
* LayoutTests/fast/transforms/non-visual-mode.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::GraphicsLayer):
(WebCore::GraphicsLayer::paintGraphicsLayerContents):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::nonVisualModeEnabled const):
(WebCore::GraphicsLayer::setNonVisualModeEnabled):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences _setNonVisualModeEnabled:]):
(-[WKPreferences _nonVisualModeEnabled]):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::createGraphicsLayer):
